### ISSUE TITLE
[Feature Request] Force Write-once / Non-updatable Updates for some users

### ISSUE THREAD

**Issue by trenta3** [HelloZeroNet/ZeroNet#1793]
It would be nice to support in `content.json` a way to specify that some IDs are allowed to send updates / write files to that location but only once, i.e. peers should ignore future updates of that files coming from said IDs.

This could be useful:
- in a write-once site, where users cannot censor themselves.
- in zerowiki, where small updates could be saved as diff wrt others, without having to trust the original user to not delete its edit.
- Implementing it in the syntax of content.json could also restrict the admin to change some part of a site after they have been written (also by having the content.json itself be subject to such a policy) which in turn could allow to create non-censorable sites (also by the owner himself) that would act as append-only logs.

In particular the last use could at least in theory solve #1288 and could partially solve some need for the more difficult site archival feature, since this one seems to be only a couple of line change.

**Comment by HelloZeroNet** [comment]
For the second usecase you described it requires append-only updates as write-once updates would allow only to post one article per user.

**Comment by trenta3** [comment]
@HelloZeroNet Yes, I stated this as write-once updates since it is more easily done and is functionally equivalent to having append-only updates, since sites could specify to source db rows from `"data-.*\.json"` instead of the current `"data\.json"`.

Of course having (also) append-only updates better suits this need, and can be as easily implemented checking if the diff updates are only adding lines, but I think it could not work in some cases:
For example, when the diff is too large (or other peers do not have that specified version) and therefore ZN fallbacks on sending the whole file update.

  I think that the receiving peer, after having checked the sign of the update, should by itself compute the diff between the two versions and only accept if the diff shows only char/lines being appended.
  Checking the sign before diffing has to be done in order not to be subject to special DoS attacks (since [diff computation is worst-case quadratic time according to difflib](https://docs.python.org/2/library/difflib.html)). 

My proposal is to wrap all of this update-check logic into a plugin to separate it from the core and make it more easily "tunable" in the future. For example, one might want to specify some sort of ACL inside the content.json that separates the insertion of records from the deletion (always using diff sequences).

**Comment by HelloZeroNet** [comment]
I try to avoid creating even more files, as it would take more time to sync and rebuilding the .db file is already taking way too much time.

> My proposal is to wrap all of this update-check logic into a plugin to separate it from the core and make it more easily "tunable" in the future. 

I would avoid this as snycing/publising between clients would be less efficient if different client would have different data stored.

**Comment by trenta3** [comment]
Ok I hadn't thought about these efficiency issues. I think the only valid idea is the one that allows only append updates to json files. When you receive an update, just diff it with what you have and if it only appends lines, then it is ok.

Another (and possibly more appropriate) way would be to parse the json file and allow it only if it adds some objects inside a key.
Since most json files are written as:
```
{
  "topics": [
    { ... },
    { ... }
  ]
}
```
this would have the same effect without having to deal with "lines" which are a fragile concept.

I think that we could also add both and see what is the better one, since I think there could be some usecases where the line-based will still be preferred.

**Comment by HelloZeroNet** [comment]
The best would be skip the json files and sync rows directly.

**Comment by trenta3** [comment]
I'm partially against that: having the data in the filesystem in simple formats is more human readable and understandable that having only a database file with everything inside.
Moreover, I've seen many database corruption since I've started using ZeroNet. What would happen when it happens?

Also possibly the only way to sync rows is to have an SQL database implementation in python and sneak in some hooks on rows update and insert. This is a point in favour since it would delete the asymmetry between inserting rows and reading the database, but I think proper care must be taken to prevent data loss by database corruption.
